### ISSUE TITLE
opt: ORDER BY inside window aggregate is ignored

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#43133]
```
> create table ab (a int, b int);
> insert into ab values (1, 10), (1, 5), (1, 20), (3, 30), (2, 20), (3, 10), (2, 10);
> select json_agg(json_build_object('a', a, 'b', b) order by json_build_object('a', a, 'b', b)) over (partition by a) from ab;
                          json_agg                          
+----------------------------------------------------------+
  [{"a": 1, "b": 10}, {"a": 1, "b": 5}, {"a": 1, "b": 20}]  
  [{"a": 1, "b": 10}, {"a": 1, "b": 5}, {"a": 1, "b": 20}]  
  [{"a": 1, "b": 10}, {"a": 1, "b": 5}, {"a": 1, "b": 20}]  
  [{"a": 2, "b": 20}, {"a": 2, "b": 10}]                    
  [{"a": 2, "b": 20}, {"a": 2, "b": 10}]                    
  [{"a": 3, "b": 30}, {"a": 3, "b": 10}]                    
  [{"a": 3, "b": 30}, {"a": 3, "b": 10}]                    
```
The `order by` is ignored here.

Note that the similar case that doesn't use windowing but uses `group by` instead works:
```
> select json_agg(json_build_object('a', a, 'b', b) order by json_build_object('a', a, 'b', b)) from ab group by a;
                          json_agg                          
+----------------------------------------------------------+
  [{"a": 1, "b": 5}, {"a": 1, "b": 10}, {"a": 1, "b": 20}]  
  [{"a": 2, "b": 10}, {"a": 2, "b": 20}]                    
  [{"a": 3, "b": 10}, {"a": 3, "b": 30}]                    
```

These are the plans for the two cases:
```
> explain (opt) select json_agg(json_build_object('a', a, 'b', b) order by json_build_object('a', a, 'b', b)) over (partition by
 a) from ab;
                          text                           
+-------------------------------------------------------+
  project                                                
   └── window partition=(1)                              
        ├── project                                      
        │    ├── scan ab                                 
        │    └── projections                             
        │         └── json_build_object('a', a, 'b', b)  
        └── windows                                      
             └── json-agg                                
                  └── variable: json_agg_1_arg1          


> explain (opt) select json_agg(json_build_object('a', a, 'b', b) order by json_build_object('a', a, 'b', b)) from ab group by a
;
                                 text                                  
+---------------------------------------------------------------------+
  project                                                              
   └── group-by                                                        
        ├── window partition=(1) ordering=+4 opt(1)                    
        │    ├── project                                               
        │    │    ├── scan ab                                          
        │    │    └── projections                                      
        │    │         └── json_build_object('a', a, 'b', b)           
        │    └── windows                                               
        │         └── windows-item: range from unbounded to unbounded  
        │              └── json-agg                                    
        │                   └── variable: column4                      
        └── aggregations                                               
             └── const-agg                                             
                  └── variable: json_agg                               

```

**Comment by RaduBerinde** [comment]
@justinj can you take a look at this? It was reported in the community slack.

**Comment by RaduBerinde** [comment]
Oh, I just realized that in this case we should put the `ORDER BY` inside the `OVER` clause. But we should at least error out if `ORDER BY` is used like this.

**Comment by RaduBerinde** [comment]
Postgres does error out in this case. We should do the same.
```
radu=> select array_agg(b order by a) over (partition by a) from ab;
ERROR:  aggregate ORDER BY is not implemented for window functions
LINE 1: select array_agg(b order by a) over (partition by a) from ab...
```

**Comment by RaduBerinde** [comment]
There is a caveat here - if we add `ORDER BY` in the windowing clause, it changes the frame. I think the right thing would be to add `RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING` to keep the frame the same.

I filed https://github.com/cockroachdb/docs/issues/6152 to document this "workaround".

**Comment by namibj** [comment]
> Oh, I just realized that in this case we should put the `ORDER BY` inside the `OVER` clause. But we should at least error out if `ORDER BY` is used like above.

I would ask you to implement the [order by argument for *_agg](https://www.cockroachlabs.com/docs/v19.2/select-clause.html#order-aggregate-function-input-rows-by-column) as a concise alternative to moving the `ORDER BY` inside the [window_specification](https://www.cockroachlabs.com/docs/v19.2/sql-grammar.html#window_specification) and adding the correct [opt_frame_clause](https://www.cockroachlabs.com/docs/v19.2/sql-grammar.html#opt_frame_clause), which refuses to work as soon as the `opt_frame_clause` or `opt_sort_clause` is specified.

It would be great to have this concise syntax, considering it suffices for many cases and is arguably more confusing if it doesn't work.

**Comment by RaduBerinde** [comment]
> refuses to work as soon as the opt_frame_clause or opt_sort_clause is specified.

If the frame is already fully specified, adding the `ORDER BY` should work fine.

If there is another, different sorting clause - well, this is actually the hard part about implementing `ORDER BY` inside aggregation. Our execution operators would need to be more powerful than they are today - they would need to be able to resort the input arguments for each aggregation, rather than relying on the windowing order. Given that postgres doesn't implement this either (and thus there are no compatibility concerns for existing postgres apps), and so far this is the only usecase we've seen (with a reasonable workaround), implementing this won't be a high priority at this time.

It seems though that you're suggesting supporting it only partially, e.g. only when there is no existing window ordering? That is doable but also a bit problematic, features that only work in certain cases can become confusing.

I agree that it is very confusing that it is ignored, we need to error out unless/until we support it.

**Comment by namibj** [comment]
> > refuses to work as soon as the opt_frame_clause or opt_sort_clause is specified.
> 
> If the frame is already fully specified, adding the `ORDER BY` should work fine.
I mean an `ORDER BY` inside the aggregate, not the window clause.
> 
> If there is another, different sorting clause - well, this is actually the hard part about implementing `ORDER BY` inside aggregation. Our execution operators would need to be more powerful than they are today - they would need to be able to resort the input arguments for each aggregation, rather than relying on the windowing order. Given that postgres doesn't implement this either (and thus there are no compatibility concerns for existing postgres apps), and so far this is the only usecase we've seen (with a reasonable workaround), implementing this won't be a high priority at this time.
I don't see this as an issue if it spits out a message about needing the verbose syntax for non-trivial cases.
> 
> It seems though that you're suggesting supporting it only partially, e.g. only when there is no existing window ordering? That is doable but also a bit problematic, features that only work in certain cases can become confusing.
I suggest it to do an AST rewrite if this is clear and trivial (neither frame nor order being specified int he window definition), while erroring out when a frame or order is already defined. It would then refer the user to the documentation or possibly provide in-line guidance.
> 
> I agree that it is very confusing that it is ignored, we need to error out unless/until we support it.
I would keep the error for all non-trivially solvable cases. I too don't think it'd be worthwhile to implement and support complex machinery for this.

To recap: I suggest implementing this as an AST-rewrite that applies the workaround behind the scenes, while throwing an error if this isn't a trivial modification.

**Comment by lukaseder** [comment]
I ran into this as well, with this reproducer:

```sql
select
  array_agg(c order by a) over (),
  array_agg(c order by a desc) over ()
from (
  values 
    (1, 1, 1),
    (2, 1, 1),
    (3, 2, 4),
    (4, 2, 2)
) as t (a, b, c)
order by a;
```

Resulting in:

```
|array_agg|array_agg|
|---------|---------|
|{1,1,4,2}|{1,1,4,2}|
|{1,1,4,2}|{1,1,4,2}|
|{1,1,4,2}|{1,1,4,2}|
|{1,1,4,2}|{1,1,4,2}|
```

The syntax should either be implemented (as in H2) or an error should be raised as in PostgreSQL:

> SQL Error [0A000]: ERROR: aggregate ORDER BY is not implemented for window functions

or DuckDB:

> Parser Error: ORDER BY is not implemented for window functions!

**Comment by lukaseder** [comment]
I've worked around this on my end. This:

```sql
array_agg(c order by a) over ()
```

Seems to be equivalent to this:

```sql
array_agg(c) over (order by a range between unbounded preceding and unbounded following)
```

It's not formally equivalent, but it appears that all SQL dialects that I've tested do not randomly re-order a previously ordered window to produce an unexpected result.
